wandb/sdk/artifacts/storage_handlers/s3_handler.py

```diff
@@ -104,39 +104,42 @@ def load_path(
         version = manifest_entry.extra.get("versionID")
 
         extra_args = {}
-        if version is None:
-            # We don't have version information so just get the latest version
-            # and fallback to listing all versions if we don't have a match.
+        if version:
+            obj = self._s3.ObjectVersion(bucket, key, version).Object()
+            extra_args["VersionId"] = version
+        else:
             obj = self._s3.Object(bucket, key)
+
+        try:
             etag = self._etag_from_obj(obj)
-            if etag != manifest_entry.digest:
-                if self.versioning_enabled(bucket):
-                    # Fallback to listing versions
-                    obj = None
-                    object_versions = self._s3.Bucket(bucket).object_versions.filter(
-                        Prefix=key
-                    )
-                    for object_version in object_versions:
-                        if (
-                            manifest_entry.extra.get("etag")
-                            == object_version.e_tag[1:-1]
-                        ):
-                            obj = object_version.Object()
-                            extra_args["VersionId"] = object_version.version_id
-                            break
-                    if obj is None:
-                        raise ValueError(
-                            "Couldn't find object version for {}/{} matching etag {}".format(
-                                bucket, key, manifest_entry.extra.get("etag")
-                            )
-                        )
-                else:
-                    raise ValueError(
-                        f"Digest mismatch for object {manifest_entry.ref}: expected {manifest_entry.digest} but found {etag}"
+        except self._botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "404":
+                raise FileNotFoundError(
+                    f"Unable to find {manifest_entry.path} at s3://{bucket}/{key}"
+                ) from e
+            raise
+
+        if etag != manifest_entry.digest:
+            # Try to match the etag with some other version.
+            if version or not self.versioning_enabled(bucket):
+                raise ValueError(
+                    f"Digest mismatch for object {manifest_entry.ref}: expected {manifest_entry.digest} but found {etag}"
+                )
+            obj = None
+            object_versions = self._s3.Bucket(bucket).object_versions.filter(Prefix=key)
+            for object_version in object_versions:
+                if manifest_entry.extra.get("etag") == self._etag_from_obj(
+                    object_version
+                ):
+                    obj = object_version.Object()
+                    extra_args["VersionId"] = object_version.version_id
+                    break
+            if obj is None:
+                raise ValueError(
```

Comment:
This should be `FileNotFoundError` as well, no?

Reply:
I think so. I was preserving the behavior that normally happens, but this also counts as a missing reference rather than as a client error.